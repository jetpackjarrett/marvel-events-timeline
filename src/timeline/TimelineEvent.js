import React, { PropTypes, Component } from 'react';
import { getImageUrl } from '../services/marvelApi';
import './timelineEvent.css';

const MONTH_NAMES = [
  'January', 'February', 'March', 'April', 'May', 'June',
  'July', 'August', 'September', 'October', 'November', 'December'
];

export default class TimelineEvent extends Component {

  static propTypes = {
    handleClick: PropTypes.func.isRequired,
    start: PropTypes.string.isRequired,
    thumbnail: PropTypes.shape({
      extension: PropTypes.string.isRequired,
      path: PropTypes.string.isRequired,
    }).isRequired,
    title: PropTypes.string.isRequired,
  };

  state = {
    imageUrl: null
  };

  componentDidMount() {
    const imageUrl = getImageUrl(this.props.thumbnail, 'standard_fantastic');
    const image = new Image();
    image.src = imageUrl;
    image.onload = () => this.setState({ imageUrl });
  }

  formatDate(date) {
    const parsed = new Date(date);
    return `${MONTH_NAMES[parsed.getMonth()]} ${parsed.getFullYear()}`;
  }

  render() {
    const { title, start, handleClick } = this.props;
    const { imageUrl } = this.state;
    const classNames = ['timeline-event'];

    if (imageUrl) {
      classNames.push('ready');
    }

    return (
      <div
        className={classNames.join(' ')}
        onClick={handleClick}
        style={{ backgroundImage: `url(${imageUrl})` }}>
        <aside>
          <h3>{title}</h3>
          <p>
            <time>{this.formatDate(start)}</time>
          </p>
        </aside>
      </div>
    );
  }
}
