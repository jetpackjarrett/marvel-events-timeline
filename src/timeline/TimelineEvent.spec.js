/*eslint import/no-extraneous-dependencies: ["error", { "devDependencies": true }] */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import TimelineEvent from './TimelineEvent';

jest.mock('../services/marvelApi');

describe('TimelineEvent', () => {
  it('renders the timeline event', () => {
    const component = shallow(
      <TimelineEvent
        handleClick={jest.fn()}
        start="1984-20-08"
        thumbnail={{
          path: 'http://fake-image.com/large',
          extension: 'jpg',
        }}
        title="Test Wars" />
    );
    component.setState({ imageUrl: 'http://fake-image.com/large.jpg' });
    expect(toJson(component)).toMatchSnapshot();
  });
});
