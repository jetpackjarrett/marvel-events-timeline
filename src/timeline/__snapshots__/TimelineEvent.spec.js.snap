exports[`TimelineEvent renders the timeline event 1`] = `
<div
  className="timeline-event ready"
  onClick={[Function]}
  style={
    Object {
      "backgroundImage": "url(http://fake-image.com/large.jpg)",
    }
  }>
  <aside>
    <h3>
      Test Wars
    </h3>
    <p>
      <time>
        undefined NaN
      </time>
    </p>
  </aside>
</div>
`;
